### PR TITLE
doc: clarify topic branches

### PR DESCRIPTION
**Pull request by nashif**
Clarify topic branches and the need for TSC approval.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

### CODE REVIEW

**Comment by fabiobaltieri** [comment]
I think topic branch description in the documentation is just confusing right now.

I remember reading that when I initially approached the project, then trying to figure out what topic branches are there the middle of the (hunderd) backport branches and what I was supposed to use, just to figure out that they were all stale and only the ble folks were using one.

Honestly I would just drop that page entirely, unless we start using topic branches for good.

**Comment by nashif** [comment]
> Honestly I would just drop that page entirely, unless we start using topic branches for good.

I am fine with that too.

**Comment by fabiobaltieri** [comment]
> At first I was like okay yes drop it would be the best thing to do, but I think it is good to have it still written down in the case if someone pops up with a topic branch in a year or so and we need to describe it again and start the discussion over.

We could always point to the old documentation. But also there's no point inviting that to happen with wording like "allows teams to work independently on a subsystem or a feature, improves efficiency and turnaround time, and encourages collaboration and streamlines communication between developers" while we are not really doing it.

The whole topic strategy has to be redefined anyway, and if I had to guess we'd rather move to topic repository more than branches (even though we are not Linux).

**Comment by kartben** [comment]
> > Honestly I would just drop that page entirely, unless we start using topic branches for good.
> 
> I am fine with that too.

+1.
Note that this page might require updating too if "shared" topic branches are officially not to be used? https://docs.zephyrproject.org/latest/contribute/guidelines.html#contribution-workflow  ("Some Zephyr subsystems do development work on a separate branch from main so you may need to indicate this in your checkout ...")

**Comment by simhein** [comment]
> > At first I was like okay yes drop it would be the best thing to do, but I think it is good to have it still written down in the case if someone pops up with a topic branch in a year or so and we need to describe it again and start the discussion over.
> 
> We could always point to the old documentation. But also there's no point inviting that to happen with wording like "allows teams to work independently on a subsystem or a feature, improves efficiency and turnaround time, and encourages collaboration and streamlines communication between developers" while we are not really doing it.
> 
> The whole topic strategy has to be redefined anyway, and if I had to guess we'd rather move to topic repository more than branches (even though we are not Linux).

Ok Thats true, didn't thought about that from that perspective but you are right.

In this case:
+1 for dropping it.

**Comment by henrikbrixandersen** [comment]
+1 for dropping the entire paragraph on topic branches.

**Comment by teburd** [comment]
> > Honestly I would just drop that page entirely, unless we start using topic branches for good.
> 
> I am fine with that too.

Please do

**Comment by mbolivar-nordic** [comment]
Process WG:

- add requirements to introduce timelines and exit criteria
- open separate issue for dedicated releases on top of LTS releases

-------------------------------------------------------------------

- @nashif: for historical context, topic branches were for short-term feature development. It wasn't clear if TSC approval is required for their creation.
- @nashif first question: do we want to continue to have topic branches at all? If so, how far do we want to go?
- @yperess when you say "when everything is ready, you create a pull request to main", does that mean as-is, or with cleaned up commits?
- @nashif there was iterative development into the topic branch, and then you have to rebase the topic branch onto upstream and clean it up, like any other pull request. The difference is you're using a topic branch in the zephyr repo instead of your fork.
- @galak I think there's value in having them, but I think it's rare. I think pushing the majority of people to sharing elsewhere on GitHub is better -- you can collaborate on someone's PR branch from a fork if they allow it.
- @mbolivar-nordic @cfriedt mentioned being able to fetch from the main zephyr repo as being a valuable thing in corporate environments that restrict where you can fetch from, which as a practical matter is something we can help out by hosting ourselves
- @galak if that's the case, there's probably some ways we could handle this -- either a repository under zephyrproject-rtos for such things, or something else, especially since there's heartache associated with topic branches
- @MaureenHelm is the motivation for this around basing releases on that? That would be different than than basing development on. Interested in knowing more about that
- @nashif this goes back to the purpose of topic branches -- purpose is for shared development of new features that are eventually deleted, not to let people release products based on topic branches which might then get deleted. Need to establish the lifecycle
- @keith-zephyr since these require TSC approval, we could require a proposed timeline along with the proposal, to be re-evaluated if the timeline expires.
- @dleach02 if the desire is for them to be short-lived, we could also define exit criteria
- @nashif another case to consider is people who want to share a new feature on top of an LTS. What about creating a release for that? E.g. having a v2.8 on top of v2.7 with additional features.
- @mbolivar-nordic interesting idea -- that would imply a formally nominated release manager and following our ongoing release process
- @nashif another point is that new features in old releases require security and safety review from @simhein and @ceolin 
- @nashif maybe need to create a separate  issue for this for things like RISC-V support based on LTS
- @gregshue one thing I struggle with when building a product over a 10 year time span based on Zephyr is I have to migrate across multiple LTS releases due to the LTS timespans. That minimizes my incentive to feature-enhance LTS; I might as well jump releases. What is the business situation for someone to have the open source project do this, instead of doing it themselves?
- @mbolivar-nordic my understanding from @cfriedt is desire to add risc-v SMP fixes in a collaborative way across multiple companies, during the release lifetime of the existing LTS
- @gregshue I can see filling in arch features into LTS as a maintenance thing
- @galak I think in the "2.8" proposal, the idea of managing additional features makes sense as long as you're willing to contribute to the release machinery. I don't think it makes sense to limit what someone would put into an LTS+additional features release
- @nashif one of my main concerns with adding new code is I still consider it a feature, even if it's architecture specific code e.g. what about adding 64-bit MIPS, MMU for Xtensa, .... I would be put off as a customer by an update to a stable product that included things like this.

**Comment by mbolivar-nordic** [comment]
Process WG: 

- pending updates to add timelines discussed above
- updates regarding governance will be made as well

------------------------------------

- @henrikbrixandersen I would like to clarify governance. I'm seeing force pushes in the risc-v topic branch
- @stephanosio one responsibility for a topic branch is periodic rebasing, so force pushing in itself isn't a problem
- @nashif this is a logistical step, not a way to add additional changes on existing branch

**Comment by cfriedt** [comment]
Sorry for joining this conversation so late - busy days / busy nights / busy weekends.

> * add requirements to introduce timelines and exit criteria

I'd be happy to support adding timelines and exit criteria if we need this specifically for the `topic-v2.7-riscv`. Should I join the process WG call next time?

> * open separate issue for dedicated releases on top of LTS releases

My preference is to not do a dedicated release on top of the LTS release, but I don't want to be the bad sheep.

**Comment by mbolivar-nordic** [comment]
Process WG:

- agree to continue refining this process
- change name from "topic" to "collab"
- interested parties: please continue to review this PR

File: doc/project/code_flow.rst
Comment (fabiobaltieri) [inline doc/project/code_flow.rst:40]: Collab?

File: doc/project/code_flow.rst
Comment (fabiobaltieri) [inline doc/project/code_flow.rst:44]: s/collabb/collab/

**Review by kartben — DISMISSED** [review]

**Review by fabiobaltieri — DISMISSED** [review]

File: doc/project/code_flow.rst
Comment (marc-hb) [inline doc/project/code_flow.rst:50]: I would drop this sentence because "fork" is an overloaded and confusing term. Here it is used to describe the intention of (not) merging back. But "fork" in the Github sense is a purely technical term, it's only about just branching in a different clone to avoid tedious and error-prone permissions and namespaces. A _Github_ "fork" is neutral and says nothing about lifecycles and future intentions, the owner of a Github fork is free to use it in any way they want. It's also the opposite of a collaboration branch, can't be both at the same time.

The previous text is very clear and enough, this last sentence just muddies the water for people not experienced with the subtle and contradictory nuances of git and Github jargons.

**Review by carlescufi — DISMISSED** [review]

File: doc/project/code_flow.rst
Comment (mbolivar-nordic) [inline doc/project/code_flow.rst:9]: No, this is done on purpose. The convention that David Kinder established is that "the zephyr repository" means https://github.com/zephyrproject-rtos/zephyr. If we use uppercase ("Zephyr"), we are referring to the project more widely. Please remove this change and do not introduce more inconsistencies. CC @kartben please take note.

File: doc/project/code_flow.rst
Comment (mbolivar-nordic) [inline doc/project/code_flow.rst:16]: s/into main/into the main/

File: doc/project/code_flow.rst
Comment (mbolivar-nordic) [inline doc/project/code_flow.rst:17]: s/a TSC/TSC/

File: doc/project/code_flow.rst
Comment (mbolivar-nordic) [inline doc/project/code_flow.rst:20]: E.g. this "Zephyr" is OK because a released version also refers to many non-zephyr repositories indirectly via zephyr/west.yml.

File: doc/project/code_flow.rst
Comment (mbolivar-nordic) [inline doc/project/code_flow.rst:48]: Everything after the comma is unclear to me

File: doc/project/code_flow.rst
Comment (mbolivar-nordic) [inline doc/project/code_flow.rst:27]: The PR is inconsistently using "Collaboration" (capitalized) and not "collaboration"; please pick one and be consistent

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]

**Review by mbolivar-nordic — DISMISSED** [review]

File: doc/project/code_flow.rst
Comment (keith-zephyr) [inline doc/project/code_flow.rst:48]:
```suggestion

```

**Review by keith-zephyr — CHANGES_REQUESTED** [review]

**Review by keith-zephyr — APPROVED** [review]

**Review by mbolivar-nordic — APPROVED** [review]

**Review by henrikbrixandersen — APPROVED** [review]

**Review by marc-hb — APPROVED** [review]

**Review by simhein — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]

**Review by stephanosio — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]

**Pull request merged**
